File: quartz/src/main/java/com/vitorpamplona/quartz/nip03Timestamp/ots/op/OpUnary.java
package com.vitorpamplona.quartz.nip03Timestamp.ots.op;

import android.util.Log;

import com.vitorpamplona.quartz.nip03Timestamp.ots.StreamDeserializationContext;

/**
 * Operations that act on a single message.
 *
 * @see Op
 */
public abstract class OpUnary extends Op {

    @Override
    public String _TAG_NAME() {
        return "";
    }

    public OpUnary() {
        super();
    }

    public static Op deserializeFromTag(StreamDeserializationContext ctx, byte tag) {
        if (tag == OpSHA1._TAG) {
            return new OpSHA1();
        } else if (tag == OpSHA256._TAG) {
            return new OpSHA256();
        } else if (tag == OpRIPEMD160._TAG) {
            return new OpRIPEMD160();
        } else if (tag == OpKECCAK256._TAG) {
            return new OpKECCAK256();
        } else {
            Log.e("OpenTimestamp", "Unknown operation tag: " + tag);

            return null;     // TODO: Is this OK? Won't it blow up later? Better to throw?
        }
    }

    @Override
    public String toString() {
        return this._TAG_NAME();
    }
}

File: quartz/src/commonMain/kotlin/com/vitorpamplona/quartz/nip03Timestamp/ots/op/OpUnary.kt
/*
 * Copyright (c) 2025 Vitor Pamplona
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.vitorpamplona.quartz.nip03Timestamp.ots.op

import com.vitorpamplona.quartz.nip03Timestamp.ots.StreamDeserializationContext
import com.vitorpamplona.quartz.utils.Log

/**
 * Operations that act on a single message.
 *
 * @see Op
 */
abstract class OpUnary : Op() {
    override fun tagName(): String = ""

    override fun toString(): String = tagName()

    companion object {
        fun deserializeFromTag(
            ctx: StreamDeserializationContext,
            tag: Byte,
        ): Op? =
            when (tag) {
                OpSHA1.TAG -> OpSHA1()
                OpSHA256.TAG -> OpSHA256()
                OpRIPEMD160.TAG -> OpRIPEMD160()
                OpKECCAK256.TAG -> OpKECCAK256()
                else -> {
                    Log.e("OpenTimestamp", "Unknown operation tag: $tag")

                    null // TODO: Is this OK? Won't it blow up later? Better to throw?
                }
            }
    }
}
